File: app/src/main/java/com/jakebarnby/filemanager/ui/adapters/SourceUsageAdapter.java
package com.jakebarnby.filemanager.ui.adapters;

import android.animation.ValueAnimator;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.dinuscxj.progressbar.CircleProgressBar;
import com.jakebarnby.filemanager.R;
import com.jakebarnby.filemanager.sources.models.Source;
import com.jakebarnby.filemanager.util.Constants;

import java.util.List;
import java.util.Locale;

/**
 * Created by Jake on 7/31/2017.
 */

public class SourceUsageAdapter extends RecyclerView.Adapter<SourceUsageAdapter.UsageViewHolder> {

    private List<Source> mSources;

    public SourceUsageAdapter(List<Source> sources) {
        mSources = sources;
    }

    @Override
    public UsageViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View inflatedView = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_usage_list, parent, false);
        return new UsageViewHolder(inflatedView);
    }

    @Override
    public void onBindViewHolder(UsageViewHolder holder, int position) {
        holder.bindHolder(mSources.get(position));
    }

    @Override
    public int getItemCount() {
        return mSources.size();
    }

    static class UsageViewHolder extends RecyclerView.ViewHolder {

        private CircleProgressBar   mPercentBar;
        private TextView            mSourceName;
        private TextView            mSourceUsage;

        UsageViewHolder(View itemView) {
            super(itemView);
            mPercentBar     = itemView.findViewById(R.id.prg_usage);
            mSourceName     = itemView.findViewById(R.id.txt_source_title);
            mSourceUsage    = itemView.findViewById(R.id.txt_space_consumption);
        }

        void bindHolder(Source source) {
            mSourceName.setText(String.format("%s%s", source.getSourceName().substring(0, 1), source.getSourceName().substring(1).toLowerCase()));
            mSourceUsage.setText(constructUsageString(source));
            simulateProgress(mPercentBar, source.getUsedSpacePercent());
        }

        private String constructUsageString(Source source) {
            double usedGb = source.getUsedSpaceGB();
            double totalGb = source.getTotalSpaceGB();

            return String.format(
                    Locale.getDefault(),
                    "%.2f / %.2f GB",
                    usedGb,
                    totalGb);
        }

        private void simulateProgress(ProgressBar bar, int maxPercent) {
            ValueAnimator animator = ValueAnimator.ofInt(0, maxPercent);
            animator.addUpdateListener(animation -> {
                int progress = (int) animation.getAnimatedValue();
                bar.setProgress(progress);

            });
            animator.setDuration(Constants.Animation.PROGRESS_DURATION);
            animator.start();
        }
    }
}

File: app/src/main/java/com/jakebarnby/filemanager/ui/adapters/SourceUsageAdapter.kt
package com.jakebarnby.filemanager.ui.adapters

import android.animation.ValueAnimator
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.dinuscxj.progressbar.CircleProgressBar
import com.jakebarnby.filemanager.R
import com.jakebarnby.filemanager.sources.models.Source
import com.jakebarnby.filemanager.ui.adapters.SourceUsageAdapter.UsageViewHolder
import com.jakebarnby.filemanager.util.Constants
import java.util.*

/**
 * Created by Jake on 7/31/2017.
 */
class SourceUsageAdapter(
    private val sources: List<Source>
) : RecyclerView.Adapter<UsageViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): UsageViewHolder {
        val inflatedView = LayoutInflater.from(parent.context)
            .inflate(R.layout.view_usage_list, parent, false)

        return UsageViewHolder(inflatedView)
    }

    override fun onBindViewHolder(holder: UsageViewHolder, position: Int) {
        holder.bindHolder(sources[position])
    }

    override fun getItemCount(): Int {
        return sources.size
    }

    class UsageViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        private val percentBar: CircleProgressBar = itemView.findViewById(R.id.prg_usage)
        private val sourceName: TextView = itemView.findViewById(R.id.txt_source_title)
        private val sourceUsage: TextView = itemView.findViewById(R.id.txt_space_consumption)

        fun bindHolder(source: Source) {
            sourceName.text = String.format("%s", source.sourceName)
            sourceUsage.text = constructUsageString(source)
            animateProgress(percentBar, source.usedSpacePercent ?: 0)
        }

        private fun constructUsageString(source: Source?): String {
            val usedGb = source?.usedSpaceGB ?: 0.0
            val totalGb = source?.totalSpaceGB ?: 0.0

            return String.format(
                Locale.getDefault(),
                "%.2f / %.2f GB",
                usedGb,
                totalGb
            )
        }

        private fun animateProgress(bar: ProgressBar, maxPercent: Int) =
            ValueAnimator.ofInt(0, maxPercent).apply {
                addUpdateListener { animation: ValueAnimator ->
                    val progress = animation.animatedValue as Int
                    bar.progress = progress
                }
                duration = Constants.Animation.PROGRESS_DURATION.toLong()

                start()
            }
    }

}